doc/ni_by_example_3.md

```diff
@@ -1,243 +1,356 @@
 
-##Perl for `ni` fundamentals
+* `my $x` indicates that `$x` is a scalar, for example a string or a number
+* `my @x` indicates that `@x` is an array
+* `my %x` indicates that `%x` is a hash
 
+What's even cooler is that, because the syntax for all of these is different, we can use all of these:
 
+```
+print $x;       //this gets the scalar value of $x
+print $x[3];    //this gets a scalar value from @x
+print $x{foo};  //this gets a scalar value from %x
+```
 
-##Connecting `bash` and `ni`
+At first glance, this is very confusing; all of these values start with `$x`--but note that the calling syntax is different for all three; you  get a scalar value (i.e `$`) out of a hash with curly braces, you get a scalar value out of an array with square brackets, and without either of those, Perl knows that you are referring to the scalar value `$x`. The syntax is a little complicated, but it's not tricky.
 
-* `e[<script>]`: Evaluate script
-  * evaluate `<script>` in bash, and stream out the results one line at a time.
+####Barewords are strings
+Consider the following `ni` spell:
 
+```
+$ ni n3p'r a, one'
+1       one
+2       one
+3       one
+(END)
+```
 
+Whereas in almost any other language, a syntax error or name error would be raised on referencing a variable that does not exist,  Perl gives the programmer a great deal of freedom to be concise. The bareword (a Perl term for a variable not prefixed with a sigil) `one` has not been defined, so Perl assumes you know what you're doing and interprets it as a string. Perl assumes you are a great programmer, and in doing so, allows you to rise to the challenge.
 
-  
-##Understanding the `ni` monitor
-More details [here](monitor.md). Overall:
 
-* Negative numbers = non-rate-determining step
-* Positive numbers = rate-determining step
-* Large Positive numbers = slow step
+####Subroutines
 
+Perl is a multi-paradigm language, but the dominant paradigm for Perl within `ni` is procedural programming. Unlike the object-oriented languages with which you are likely familiar, where methods are objects (often "first-class" objects, [whatever that means](http://stackoverflow.com/questions/245192/what-are-first-class-objects)) procedural languages focus on their functions, called "subroutines."  
 
-##JSON I/O
-* `json_encode`
-* `D:<field1>,:<field2>...`: JSON Destructure
-  * `ni` implements a very fast JSON parser that is great at pulling out string and numeral fields.
-  * As of 2016-12-24, the JSON destructurer does not support list-based fields in JSON.
+Perl subroutines store the variables with which they are called in a default variable named `@_`. Take a moment here to think about how one would refer the elements in `@_`.
 
-##Directory I/O
+Subroutines are stored and referenced with the following syntax:
+
+```
+*x = sub {"hi"}
+$v = &x // $v will be set to "hi"
```

Comment:
Haha, I totally misread this line; `//` is the defined-or operator in Perl (v5.14 and later or so, IIRC) -- you want `#` to indicate the comment so perlers don't lose their minds :)